spring-integration-ip/src/main/java/org/springframework/integration/ip/config/IpAdapterParserUtils.java

```diff
@@ -150,7 +153,40 @@ public static void addConstuctorValueIfAttributeDefined(BeanDefinitionBuilder bu
 	 * @param builder The builder.
 	 * @param parserContext The parser context.
 	 */
-	public static void addHostAndPortToConstructor(Element element,
+	public static void addBasicSocketConfigToConstructor(Element element,
+			BeanDefinitionBuilder builder, ParserContext parserContext) {
+		String socketExpression = element.getAttribute(IpAdapterParserUtils.SOCKET_EXPRESSION);
+		if (StringUtils.hasText(socketExpression)) {
+			addSocketExpressionToConstructor(socketExpression, element, builder, parserContext);
+		} else {
+			addHostAndPortToConstructor(element, builder, parserContext);
+		}
+	}
+
+	private static void addSocketExpressionToConstructor(String socketExpression,
+			Element element, BeanDefinitionBuilder builder, ParserContext parserContext) {
+		String host = element.getAttribute(IpAdapterParserUtils.HOST);
+		String port = element.getAttribute(IpAdapterParserUtils.PORT);
+		String ack = element.getAttribute(IpAdapterParserUtils.ACK);
+		String multicast = IpAdapterParserUtils.getMulticast(element);
+		if (StringUtils.hasText(host) || StringUtils.hasText(port) || ack.equals("true")
+				|| multicast.equals("true")) {
```

Comment:
We can't check the actual value in the attribute from the `Parser`, just because it can be specified in the `property-placeholder`/SpEL style, e.g.:
```
multicast="${myProj.udp.multicust}"
---------
acknowledge="#{myBeanForUdp.acknowledge}"
```
Seems for me we have `Assert.state(!this.acknowledge` in case of `socketExpression` in the `UnicastSendingMessageHandler`.
The same we should do for the `ack` if that.
Here we can check only host+port/socket-expression exclusiveness.